src/Database.php

```diff
@@ -300,6 +304,59 @@ public function createCollection(string $collectionName, array $options = [])
         return $result;
     }
 
+    /**
+     * Create a new encrypted collection explicitly.
+     *
+     * This function will automatically create data keys for any encrypted
+     * fields where the "keyId" option is null. This function will return a copy
+     * of the modified "encryptedFields" option in addition to the result from
+     * createCollection(). The "encryptedFields" option is required.
+     *
+     * If any error is encountered while creating data keys or invoking
+     * createCollection(), a CreateEncryptedCollectionException will be thrown.
+     * The original exception and modified "encryptedFields" option can be
+     * accessed via the getPrevious() and getEncryptedFields() methods,
+     * respectively.
+     *
+     * @see CreateCollection::__construct() for supported options
+     * @return array A tuple consisting of the createCollection() result and modified "encryptedFields" option
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws CreateEncryptedCollectionException for any errors creating data keys or invoking createCollection()
+     */
+    public function createEncryptedCollection(string $collectionName, ClientEncryption $clientEncryption, string $kmsProvider, ?array $masterKey, array $options): array
+    {
+        if (! isset($options['encryptedFields']) || ! is_array($options['encryptedFields']) && ! is_object($options['encryptedFields'])) {
+            throw InvalidArgumentException::invalidType('"encryptedFields" option', $options['encryptedFields'] ?? null, ['array', 'object']);
+        }
+
+        $encryptedFields = (array) recursive_copy($options['encryptedFields']);
+
+        $createDataKeyArgs = [
+            $kmsProvider,
+            isset($masterKey) ? ['masterKey' => $masterKey] : [],
+        ];
+
+        try {
+            if (isset($encryptedFields['fields']) && is_array($encryptedFields['fields'])) {
+                foreach ($encryptedFields['fields'] as &$field) {
+                    if (is_array($field) && array_key_exists('keyId', $field) && $field['keyId'] === null) {
+                        $field['keyId'] = $clientEncryption->createDataKey(...$createDataKeyArgs);
+                    } elseif (is_object($field) && property_exists($field, 'keyId') && $field->keyId === null) {
+                        $field->keyId = $clientEncryption->createDataKey(...$createDataKeyArgs);
+                    }
+                }
+
+                $options['encryptedFields'] = $encryptedFields;
+            }
+
+            $result = $this->createCollection($collectionName, $options);
+
+            return [$result, $encryptedFields];
```

Comment:
Is returning `$encryptedFields` the reason why we're copying the option earlier? What's the advantage of this approach over modifying `$options['encryptedFields']` directly and returning that?

Reply:
If `$options['encryptedFields']` is an object, we risk modifying the user's input argument. This is admittedly an edge case, but we technically allow `array|object` for "document" options such as this and I didn't want to impose a restriction.

On a related note, I'm not a fan of needing separate code paths for modifying `keyId`, but the field document (i.e. value of `$options['encryptedFields'][]`) can also be `array|object`. This has started me thinking on some utility APIs that might make it more pleasant to work with arrays and stdClass objects interchangeably.

There's another edge case I intentionally didn't explore, which is a `MongoDB\BSON\Serializable` instance being used for these options. Handling that would require eagerly "serializing" the input in the helper method before we inspect and modify `$options['encryptedFields']`. I don't think it's worth handling that (very unlikely to occur in practice) but I think it may deserve a comment to acknowledge the edge case.

Reply:
Ah, yes, objects would be passed by reference and thus changed. That makes sense 👍